Controller/Payment/Notification.php

```diff
@@ -3,130 +3,286 @@
-            }
+            $order_note = $note_prefix . 'Canceled Payment - ' . $payment_type;
+            $this->cancelOrder($order, Order::STATE_CANCELED, $order_note);
         } elseif ($transaction == 'expire') {
             if ($order->canCancel()) {
                 $order_note = $note_prefix . 'Expired Payment - ' . $payment_type;
-                $this->cancelOrder($orderId, Order::STATE_CANCELED, $order_note);
+                $this->cancelOrder($order, Order::STATE_CANCELED, $order_note);
             }
         } elseif ($transaction == 'deny') {
+            $this->setPaymentInformation($order, $trxId, $payment_type);
             $order_note = $note_prefix . 'Payment Deny - ' . $payment_type;
-            $order->addStatusToHistory(Order::STATE_PAYMENT_REVIEW, $order_note, false);
+            $this->setOrderStateAndStatus($order, Order::STATE_PAYMENT_REVIEW, $order_note);
         } elseif ($transaction == 'refund' || $transaction == 'partial_refund') {
-            $isFullRefund = $transaction == 'refund';
-
-            /**
-             * Get last array object from refunds array and get the value from last refund object
-             */
-            $refunds = $notif->refunds;
-            $refund[] = end($refunds);
-            $refund_key = $refund[0]->refund_key;
-            $refund_amount = $refund[0]->refund_amount;
-            $refund_reason = $refund[0]->reason;
 
             /**
              * Do not process if the notification contain 'bank_confirmed_at' from request body
              */
-            $refundRaw[] = end($body['refunds']);
+            $refundRaw[] = end($rawBody['refunds']);
             if (isset($refundRaw[0]['bank_confirmed_at'])) {
                 return $this->getResponse()->setBody('OK');
-            }
+            } else {
+                /**
+                 * Get last array object from refunds array
+                 */
+                $refunds = $midtransStatusResult->refunds;
+                $refund[] = end($refunds);
+                $refund_reason = $refund[0]->reason;
 
-            /**
-             * Handle fullRefund: if refunded from midtrans dashboard close order and add comment history ,
-             * If refund from magento dashboard add only comment history.
-             */
-            if ($isFullRefund) {
-                $refund_note = $note_prefix . 'Full Refunded: ' . $refund_amount . '  |  Refund-Key: ' . $refund_key . '  |  Reason: ' . $refund_reason;
-                if ($order->getStatus() != Order::STATE_CLOSED || $order->getState() != Order::STATE_CLOSED && $this->canFullRefund($refund_key, $order, $refund_amount) == true) {
-                    $this->cancelOrder($orderId, Order::STATE_CLOSED, $refund_note);
+                /**
+                 * Get order-id from refund reason, this is process refund from Magento dashboard
+                 */
+                $midtransOrderId = $this->getOrderIdFromReason($refund_reason);
```

Comment:
What is the refund_reason string that is treated as `refund from Magento Dashboard`? is it `someString :order_id` ?

We need to note this down (on readme?) saying that becareful when refunding from MAP with format `someString :some_number`  because some_number can be a valid Magento order and get mistakenly auto-processed as refund. For example if merchant refunding from Map with reason `refund for customer:321`

Reply:
Alternatively you can use some less usual delimiter, instead of `:` use `:::` or `###`.

Reply:
Oh, looks like the refund_reason string will have `from Magento dashboard order :`, maybe just check if that string is exist.

Reply:
Yes if the refund from Magento dashboard will use this format
`from Magento dashboard order :VALUE-ORDER-ID`
, I used `:` as delimiter value to get the order-id from reasons. I think is good idea to use `:::` or `###` instead. 

Thank you for the insight da